add basic "howto" for using Terraform with EKS-Anywhere (#1192)
aws/eks-anywhere: examples/terraform/README.md

> ## How-To: using Terraform to manage an EKS-A Cluster

Reviewer: nit: use full name "EKS Anywhere" the first time and abbreviate thereafter
```suggestion
## How-To: using Terraform to manage an EKS Anywhere (EKS-A) Cluster
```

> A minimal guide which lays out how you can use Terraform to manage and modify an EKS-A clusters. This guide is meant for illustrative purposes only and is not meant as a definitive approach to building production systems with Terraform and EKS Anywhere.

Reviewer: ```suggestion
A minimal guide which lays out how you can use Terraform to manage and modify an EKS-A cluster. This guide is meant for illustrative purposes only and is not meant as a definitive approach to building production systems with Terraform and EKS Anywhere.
```

in the case of a stand-alone cluster, the workload cluster itself.
We can manage a subset of the fields in the EKS-A CRDs with any tool that can interact with the Kubernetes API, like `kubectl` or, in this case, the Terraform Kubneretes provider.

In this guide, we'll show you how to import your EKS-A cluster into Terraform state and 
how to scale your EKS-A worker nodes using the Terraform Kubernetes provider.

### Prerequisites
- An existing `EKS-A` cluster

- the latest version of [Terraform](https://www.terraform.io/downloads)

- the latest version of [tfk8s](https://github.com/jrhouston/tfk8s), a tool for converting K8s manifest files to Terraform HCL

### Guide
0. Create an EKS-A management cluster, or a self-managed stand-alone cluster. 
- if you already have an existing EKS-A cluster, skip this step.
- if you don't already have an existing EKS-A cluster, follow [the official instructions to create one](https://anywhere.eks.amazonaws.com/docs/getting-started/install/)

1. Convert your EKS-A cluster spec kubernetes manifest into Terraform HCL with `tfk8s`
- install [tfk8s](https://github.com/jrhouston/tfk8s#install)
```

> export MY_EKSA_CLUSER="myCLusterName"

Author: ```suggestion
export MY_EKSA_CLUSER="myClusterName"
```

kubectl get cluster ${MY_EKSA_CLUSTER} -o yaml | tfk8s --strip -o ${MY_EKSA_CLUSTER}.tf
```
Clean up the Terraform configuration generated by `tfk8s`
- remove the `generation` field in the `metadata` of the cluster
- add the `namespace` `default` to the `metadata` of the cluster
  
Your cluster manifest metadata should look like this:
```
  manifest = {
    "apiVersion" = "anywhere.eks.amazonaws.com/v1alpha1"
    "kind" = "Cluster"
    "metadata" = {
      "name" = "MyClusterName"
      "namespace" = "default"
    }
```

2. Set up the Terraform Kubernetes provider
```
export KUBECONFIG=/path/to/my/kubeconfig.kubeconfig
cat << EOF >> ./provider.tf
provider "kubernetes" {
  config_path    = "${KUBECONFIG}"
}
EOF
```

3. Import your EKS-A cluster into terraform state
```

> export MY_EKSA_CLUSER="myCLusterName"

Reviewer 2: ```suggestion
export MY_EKSA_CLUSTER="myCLusterName"
```

> export MY_EKSA_CLUSTER="myCLusterName"

Author: ```suggestion
export MY_EKSA_CLUSTER="myClusterName"
```

terraform init

> terraform import kubernetes_manifest.cluster_${MY_EKA_CLUSTER} "apiVersion=anywhere.eks.amazonaws.com/v1alpha1,kind=Cluster,namespace=default,name=${MY_EKA_CLUSTER}"

Reviewer 2: ```suggestion
terraform import kubernetes_manifest.cluster_${MY_EKSA_CLUSTER} "apiVersion=anywhere.eks.amazonaws.com/v1alpha1,kind=Cluster,namespace=default,name=${MY_EKSA_CLUSTER}"
```

```

> 4. check for differences between the state and your cluster

Reviewer: nit: capitalize
```suggestion
4. Check for differences between the state and your cluster
```

- modify the `count` value of one of your `workerNodeGroupConfigurations`, or another mutable field, in the configuration stored in `${MY_EKSA_CLUSTER}.tf` file.
- check the expected diff between your cluster state and the modified local state via `terraform plan`

> you should see in the output that the worker node group configuration count field (or whichever filed you chose to modify) will be modfied by Terraform.

Reviewer: ```suggestion
You should see in the output that the worker node group configuration count field (or whichever field you chose to modify) will be modified by Terraform.
```

> ## How-To: using Terraform to manage an EKS Anywhere (EKS-A) Cluster

Reviewer 3: ```suggestion
## Using Terraform to manage an EKS Anywhere Cluster
```
I'll try to apply our docs style to my editing here. We try to avoid "EKS-A" in the docs and only spell it out unless it is some literal value.

> A minimal guide which lays out how you can use Terraform to manage and modify an EKS-A cluster. This guide is meant for illustrative purposes only and is not meant as a definitive approach to building production systems with Terraform and EKS Anywhere.

Reviewer 3: ```suggestion
This is a minimal guide that lays out how you can use Terraform to manage and modify an EKS Anywhere cluster. The guide is meant for illustrative purposes only and is not meant as a definitive approach to building production systems with Terraform and EKS Anywhere.
```

> At its heart, EKS-A is a set of Kubernetes CRDs which define an EKS-A cluster, 

Reviewer 3: ```suggestion
At its heart, EKS Anywhere is a set of Kubernetes CRDs which define an EKS Anywhere cluster, 
```

and a controller which moves the cluster to match these definitions. 

> These CRDs, and the EKS-A controller, live on the management cluster, or, 

Reviewer 3: ```suggestion
These CRDs, and the EKS-A controller, live on the management cluster or
```

> on a self-managed cluster```

Reviewer 3: ```suggestion
on a self-managed cluster.
```

> in the case of a stand-alone cluster, the workload cluster itself.

Reviewer 4: i feel we gonna settle down a name for this "stand-alone management/workload cluster" across the board. kinda confusing to call it workload cluster (tho it is a workload cluster) and mix it up w workload cluster in management setting

Author: hm yeah that's a good point. I wasn't sure what to call it here. Maybe just 'self-managed cluster'?

> We can manage a subset of the fields in the EKS-A CRDs with any tool that can interact with the Kubernetes API, like `kubectl` or, in this case, the Terraform Kubneretes provider.

Reviewer 3: ```suggestion
We can manage a subset of the fields in the EKS Anywhere CRDs with any tool that can interact with the Kubernetes API, like `kubectl` or, in this case, the Terraform Kubernetes provider.
```

> In this guide, we'll show you how to import your EKS-A cluster into Terraform state and 

Reviewer 3: ```suggestion
In this guide, we'll show you how to import your EKS Anywhere cluster into Terraform state and 
```

> how to scale your EKS-A worker nodes using the Terraform Kubernetes provider.

Reviewer 3: ```suggestion
how to scale your EKS Anywhere worker nodes using the Terraform Kubernetes provider.
```

### Prerequisites

> - An existing `EKS-A` cluster

Reviewer 3: ```suggestion
- An existing EKS Anywhere cluster
```

- the latest version of [Terraform](https://www.terraform.io/downloads)

> - the latest version of [tfk8s](https://github.com/jrhouston/tfk8s), a tool for converting K8s manifest files to Terraform HCL

Reviewer 3: ```suggestion
- the latest version of [tfk8s](https://github.com/jrhouston/tfk8s), a tool for converting Kubernetes manifest files to Terraform HCL
```

### Guide
0. Create an EKS-A management cluster, or a self-managed stand-alone cluster. 
- if you already have an existing EKS-A cluster, skip this step.
- if you don't already have an existing EKS-A cluster, follow [the official instructions to create one](https://anywhere.eks.amazonaws.com/docs/getting-started/install/)

1. Set up the Terraform Kubernetes provider

> ```

Reviewer 3: ```suggestion
   ```bash
```

> export KUBECONFIG=/path/to/my/kubeconfig.kubeconfig

Reviewer 3: ```suggestion
   export KUBECONFIG=/path/to/my/kubeconfig.kubeconfig
```

> cat << EOF >> ./provider.tf

Reviewer 3: ```suggestion
   cat << EOF >> ./provider.tf
```

> provider "kubernetes" {

Reviewer 3: ```suggestion
   provider "kubernetes" {
```

>   config_path    = "${KUBECONFIG}"

Reviewer 3: ```suggestion
     config_path    = "${KUBECONFIG}"
```

> }

Reviewer 3: ```suggestion
   }
```

> EOF

Reviewer 3: ```suggestion
   EOF
```

> ```

Reviewer 3: ```suggestion
   ```
```

> 2. Convert your EKS-A cluster Kubernetes manifest into Terraform HCL with `tfk8s`

Reviewer 3: ```suggestion
2. Get  `tfk8s` and use it to convert your EKS Anywhere cluster Kubernetes manifest into Terraform HCL:
```

> 2. Convert your EKS-A cluster kubernetes manifest into Terraform HCL with `tfk8s`

Author: ```suggestion
2. Convert your EKS-A cluster Kubernetes manifest into Terraform HCL with `tfk8s`
```

> - install [tfk8s](https://github.com/jrhouston/tfk8s#install)

Reviewer 3: ```suggestion
   - Install [tfk8s](https://github.com/jrhouston/tfk8s#install)
```

> ```

Reviewer 3: ```suggestion
   - Convert the manifest into Terraform HCL:
   ```
```

> export MY_EKSA_CLUSTER="myClusterName"

Reviewer 3: ```suggestion
   export MY_EKSA_CLUSTER="myClusterName"
```

> export MY_EKSA_CLUSER="myClusterName"

Author: ```suggestion
export MY_EKSA_CLUSTER="myClusterName"
```

> kubectl get cluster ${MY_EKSA_CLUSTER} -o yaml | tfk8s --strip -o ${MY_EKSA_CLUSTER}.tf

Reviewer 3: ```suggestion
   kubectl get cluster ${MY_EKSA_CLUSTER} -o yaml | tfk8s --strip -o ${MY_EKSA_CLUSTER}.tf
```

> ```

Reviewer 3: ```suggestion
   ```
```

3. Configure the Terraform cluster resource definition generated in step 2

> - add the `namespace` `default` to the `metadata` of the cluster

Author: ```suggestion
- add the `namespace` `default` to the `metadata` of the cluster
- remove the `generation` field from the `metadata` of the cluster
```

> - add the `namespace` `default` to the `metadata` of the cluster

Reviewer 3: ```suggestion
   - Add the `namespace` `default` to the `metadata` of the cluster
```

> - remove the `generation` field from the `metadata` of the cluster

Reviewer 3: ```suggestion
   - Remove the `generation` field from the `metadata` of the cluster
```

> Your cluster manifest metadata should look like this (`generation` may be different):

Reviewer 3: ```suggestion
   Your cluster manifest metadata should look like this (`generation` may be different):
```

> ```

Reviewer 3: ```suggestion
   ```bash
```

  manifest = {
    "apiVersion" = "anywhere.eks.amazonaws.com/v1alpha1"
    "kind" = "Cluster"
    "metadata" = {
      "name" = "MyClusterName"
      "namespace" = "default"
    }

> ```

Reviewer 3: ```suggestion
   ```
```

> Set `metadata.generation` as a [computed field](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#computed-fields)

Reviewer 3: ```suggestion
   Set `metadata.generation` as a [computed field](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#computed-fields)
```

> - add the following to your cluster resource configuration

Reviewer 3: ```suggestion
   - Add the following to your cluster resource configuration
```

> ```

Reviewer 3: ```suggestion
   ```
```

> computed_fields=["metadata.generated"]

Author: ```suggestion
computed_fields = ["metadata.generated"]
```

> computed_fields = ["metadata.generated"]

Reviewer 3: ```suggestion
   computed_fields = ["metadata.generated"]
```

> ```

Reviewer 3: ```suggestion
   ```
```

> Configure the field manager to [force reconcile managed resources](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#field_manager)

Reviewer 3: ```suggestion
   Configure the field manager to [force reconcile managed resources](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#field_manager)
```

> - add the following configuration block to your cluster resource:

Reviewer 3: ```suggestion
   - Add the following configuration block to your cluster resource:
```

> ```

Reviewer 3: ```suggestion
   ```bash
```

>   field_manager {

Reviewer 3: ```suggestion
   field_manager {
```

>     force_conflicts = true

Reviewer 3: ```suggestion
     force_conflicts = true
```

>   }

Reviewer 3: ```suggestion
   }
```

> ```

Reviewer 3: ```suggestion
   ```
```

> 4. Import your EKS-A cluster into terraform state

Reviewer 3: ```suggestion
4. Import your EKS Anywhere cluster into terraform state:
```

> ```

Reviewer 3: ```suggestion
   ```bash
```

> export MY_EKSA_CLUSTER="myClusterName"

Reviewer 3: ```suggestion
   export MY_EKSA_CLUSTER="myClusterName"
```

> terraform init

Reviewer 3: ```suggestion
   terraform init
```

> terraform import kubernetes_manifest.cluster_${MY_EKSA_CLUSTER} "apiVersion=anywhere.eks.amazonaws.com/v1alpha1,kind=Cluster,namespace=default,name=${MY_EKSA_CLUSTER}"

Reviewer 3: ```suggestion
   terraform import kubernetes_manifest.cluster_${MY_EKSA_CLUSTER} "apiVersion=anywhere.eks.amazonaws.com/v1alpha1,kind=Cluster,namespace=default,name=${MY_EKSA_CLUSTER}"
```

> ```

Reviewer 3: ```suggestion
   ```
```

> After you `import` your cluster, you will need to run `terraform apply` one time to ensure that the `manifest` field of your cluster resource is in-sync. 

Reviewer 3: ```suggestion
   After you `import` your cluster, you will need to run `terraform apply` one time to ensure that the `manifest` field of your cluster resource is in-sync. 
```

> This will not change the state of your cluster, but is a required step after the initial import.

Reviewer 3: ```suggestion
   This will not change the state of your cluster, but is a required step after the initial import.
```

> The `manifest` field stores the contents of the associated kubernetes manifest, while the `object` field stores the actual state of the resource.

Reviewer 3: ```suggestion
   The `manifest` field stores the contents of the associated kubernetes manifest, while the `object` field stores the actual state of the resource.
```

5. Modify Your Cluster using Terraform

> - modify the `count` value of one of your `workerNodeGroupConfigurations`, or another mutable field, in the configuration stored in `${MY_EKSA_CLUSTER}.tf` file.

Reviewer 3: ```suggestion
   - Modify the `count` value of one of your `workerNodeGroupConfigurations`, or another mutable field, in the configuration stored in `${MY_EKSA_CLUSTER}.tf` file.
```

> - check the expected diff between your cluster state and the modified local state via `terraform plan`

Reviewer 3: ```suggestion
   - Check the expected diff between your cluster state and the modified local state via `terraform plan`
```

> You should see in the output that the worker node group configuration count field (or whichever field you chose to modify) will be modified by Terraform.

Reviewer 3: ```suggestion
   You should see in the output that the worker node group configuration count field (or whichever field you chose to modify) will be modified by Terraform.
```

> Now, actually change your cluster to match the local configuration:

Reviewer 3: ```suggestion
6. Now, actually change your cluster to match the local configuration:
```

> - `terraform apply`

Reviewer 3: ```suggestion
   ```bash
   terraform apply
   ```
```

> - observe the change to your cluster -- `kubectl get nodes`, for example, and observe new worker nodes being created

Reviewer 3: ```suggestion
7. Observe the change to your cluster. For example:
   ```bash
   kubectl get nodes
   ```
```
